p2p/discovery.py

```diff
@@ -820,6 +946,63 @@ def __init__(self, proto: DiscoveryProtocol, peer_pool: PeerPool,
         await self.proto.stop()
 
 
+class NodeTicketInfo:
+    last_issued: int = 0
+    last_used: int = 0
+
+
+class TopicEntry:
+
+    def __init__(self, node: kademlia.Node, expiry: float) -> None:
+        self.node = node
+        self.expiry = expiry
+
+
+class TopicTable:
+    registration_lifetime = 60 * 60
+
+    def __init__(self, logger: TraceLogger) -> None:
+        self.logger = logger
+        # A per-topic FIFO list of nodes.
+        self.topics: Dict[bytes, List[TopicEntry]] = {}
+        # The IDs of the last issued/used tickets for any given node.
+        self.node_tickets: Dict[kademlia.Node, NodeTicketInfo] = {}
+
+    def add_node(self, node: kademlia.Node, topic: bytes) -> None:
+        entries = self.topics.setdefault(topic, [])
+        if len(entries) >= MAX_ENTRIES_PER_TOPIC:
+            entries.pop(0)
```

Comment:
Thoughts on changing this to be `entries = entries[-1 * MAX_ENTRIES_PER_TOPIC:]` so that this will handle the case where it exceeds the maximum by more than 1?

Reply:
This is actually buggy as it will not pop one item when we're at max capacity, right? Went with this instead:
```
        if len(entries) >= MAX_ENTRIES_PER_TOPIC:
            # We add 1 to :to_drop: here because we want to have room for the new item we're
            # adding.
            to_drop = (len(entries) - MAX_ENTRIES_PER_TOPIC) + 1
            entries = cytoolz.drop(entries, to_drop)
```